File: algorithms/lab06/python/01-bfs/01-bfs.py
# SPDX-License-Identifier: BSD-3-Clause

import sys
from collections import deque

class Task:
    def __init__(self) -> None:
        # n = numar de noduri, m = numar de muchii/arce
        self.n = 0
        self.m = 0
        
        # adj[node] = lista de adiacenta a nodului node
        # exemplu: daca adj[node] = [..., neigh, ...] => exista muchia (node, neigh)
        self.adj = []
        
        # nodul sursa in parcurgerea BFS
        self.source = 0

        # distanta maxima in graf
        self.INF = int(1e9) # 10^9 = 1.000.000.000

    def solve(self) -> None:
        self.read_input()
        self.write_output(self.get_result())

    def read_input(self) -> None:
        with open("in", "r") as fin:
            tokens = fin.read().split()
            if not tokens:
                return
            
            self.n = int(tokens[0])
            self.m = int(tokens[1])
            self.source = int(tokens[2])
            
            self.adj = [[] for _ in range(self.n + 1)]
            
            idx = 3
            for _ in range(self.m):
                x = int(tokens[idx])
                y = int(tokens[idx+1])
                idx += 2
                self.adj[x].append(y)
                self.adj[y].append(x)

    def get_result(self) -> list[int]:
        return self.bfs(self.source)

    # Complexitate: O(n + m)
    def bfs(self, source: int) -> list[int]:
        # Step 0: alocare vector de distante
        # d[node] = INF, oricare ar fi node
        d = [self.INF] * (self.n + 1)

        # Step 1: declaram o coada in care putem baga noduri
        q = deque()

        # Step 2: pasul initial: pun in coada TOATE nodurile cu grad intern 0
        q.append(source)
        d[source] = 0

        # Step 3: cat timp NU am parcurs toate nodurile
        while q:
            # Step 3.1: extragem si procesam urmatorul nod din coada
            node = q.popleft()

            # Step 3.2: parcurgem vecinii lui node
            for neigh in self.adj[node]:
                # actulizam distanta lui x fata de sursa daca
                # am gasit o solutie mai buna
                if d[node] + 1 < d[neigh]:
                    # adauga nodul in coada si actualizeaza distanta pana la el
                    d[neigh] = d[node] + 1
                    q.append(neigh)

        # Step 4: problema ne cere ca acolo unde nu exista drum sa punem -1
        # inlocuim valorile egale cu INF din d cu -1
        for i in range(1, self.n + 1):
            if d[i] == self.INF:
                d[i] = -1

        return d

    def write_output(self, d: list[int]) -> None:
        with open("out", "w") as fout:
            out_str = " ".join(map(str, d[1:self.n+1]))
            fout.write(out_str + "\n")

def main() -> None:
    task = Task()
    task.solve()

# [ATENTIE] NU modifica functia main!
if __name__ == "__main__":
    main()

File: algorithms/lab06/rust/01-bfs/01-bfs.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::collections::VecDeque;
use std::fs::File;
use std::io::{BufRead, BufReader, Write};

#[derive(Default)]
struct Task {
    // n = numar de noduri, m = numar de muchii/arce
    n: i32,
    m: i32,

    // adj[node] = lista de adiacenta a nodului node
    // exemplu: daca adj[node] = {..., neigh, ...} => exista muchia (node, neigh)
    adj: Vec<Vec<i32>>,

    // nodul sursa in parcurgerea BFS
    source: i32,
}

impl Task {
    // numarul maxim de noduri
    const NMAX: usize = 100_005; // 10^5 + 5

    // distanta maxima in graf
    const INF: i32 = 1_000_000_000; // 10^9

    pub fn solve(&mut self) {
        self.read_input();
        self.write_output(&self.get_result());
    }

    fn read_input(&mut self) {
        let fin = File::open("in").expect("in");
        let mut lines = BufReader::new(fin).lines().map(|l| l.unwrap());
        let first = lines.next().unwrap();
        let mut tokens = first.split_whitespace().map(|s| s.parse::<i32>().unwrap());
        self.n = tokens.next().unwrap();
        self.m = tokens.next().unwrap();
        self.source = tokens.next().unwrap();
        self.adj = vec![vec![]; Self::NMAX];
        for _ in 0..self.m {
            let line = lines.next().unwrap();
            let mut tok = line.split_whitespace().map(|s| s.parse::<i32>().unwrap());
            let x = tok.next().unwrap();
            let y = tok.next().unwrap();
            // muchie (x, y)
            self.adj[x as usize].push(y);
            self.adj[y as usize].push(x);
        }
    }

    fn get_result(&self) -> Vec<i32> {
        return self.bfs(self.source);
    }

    // Complexitate: O(n + m)
    fn bfs(&self, source: i32) -> Vec<i32> {
        // Step 0: alocare vector de distante
        // d[node] = INF, oricare ar fi node
        let mut d = vec![Self::INF; (self.n + 1) as usize];

        // Step 1: declaram o coada in care putem baga noduri
        let mut q = VecDeque::new();

        // Step 2: pasul initial: pun in coada TOATE nodurile cu grad intern 0
        q.push_back(source);
        d[source as usize] = 0;

        // Step 3: cat timp NU am parcurs toate nodurile
        while let Some(node) = q.pop_front() {
            // Step 3.1: extragem si procesam urmatorul nod din coada
            // Step 3.2: parcurgem vecinii lui node
            for &neigh in &self.adj[node as usize] {
                // actualizam distanta lui neigh fata de sursa daca
                // am gasit o solutie mai buna
                if d[node as usize] + 1 < d[neigh as usize] {
                    // adauga nodul in coada si actualizeaza distanta pana la el
                    d[neigh as usize] = d[node as usize] + 1;
                    q.push_back(neigh);
                }
            }
        }

        // Step 4: problema ne cere ca acolo unde nu exista drum sa punem -1
        // inlocuim valorile egale cu INF din d cu -1
        for value in &mut d {
            if *value == Self::INF {
                *value = -1;
            }
        }

        return d;
    }

    fn write_output(&self, d: &[i32]) {
        let mut fout = File::create("out").expect("out");
        for node in 1..=self.n {
            let idx = node as usize;
            if node == self.n {
                writeln!(fout, "{}", d[idx]).unwrap();
            } else {
                write!(fout, "{} ", d[idx]).unwrap();
            }
        }
    }
}

// [ATENTIE] NU modifica functia main!
fn main() {
    let mut task = Box::new(Task::default());
    task.solve();
}
